static/opta.sh

```diff
@@ -0,0 +1,246 @@
+    AWSCLI2URL="https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip"
+    KUBECTLURL="https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/$ARCH/kubectl"
+    GOOGLEURL="https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-sdk-365.0.1-linux-x86_64.tar.gz"
+    DOCKERURL="https://download.docker.com/linux/static/stable/x86_64/docker-20.10.9.tgz"
+fi
+
+if  [ ! "$(command -v docker)" ]; then
+    echo "opta.sh requires Docker on your machine, please install docker."
+    exit 1
+fi
+
+##
+MYUID=`id -u ${USER}`
+MYGID=`id -g ${USER}`
+CONTAINERGID=$(( MYGID < 1000 ? 12345 : MYGID ))
+
+
+
+mkdir -p /tmp/opta
+
+cat << EOF > /tmp/opta/install.sh
+#! /usr/bin/env bash
+
+set -u
+
+NONINTERACTIVE=1
+
+echo "Going to install opta v$OPTAVERSION"
+
+echo $OPTAURL
+echo "Downloading installation package..."
+curl -s ${OPTAURL} -o /tmp/opta.zip --fail
+if [[ $? != 0 ]]; then
+  echo "Version $OPTAVERSION not found."
+  echo "Please check the available versions at https://github.com/run-x/opta/releases."
+  exit 1
+fi
+
+
+echo "Installing..."
+unzip -q /tmp/opta.zip -d /usr/local/bin/opta
+chmod u+x /usr/local/bin/opta/opta
+chown -R ${USER}:${USER} /usr/local/bin/opta
+
+EOF
+
+chmod +x /tmp/opta/install.sh
+
+cat << EOF > /tmp/opta/Dockerfile
+# base image
+FROM python:3.8.12-bullseye
+RUN apt-get update
+
+# aws cli
+RUN apt install unzip curl groff less -y
+RUN curl "$AWSCLI2URL" -o "awscliv2.zip"
+RUN unzip awscliv2.zip
+RUN ./aws/install
+RUN rm  awscliv2.zip
+
+# gcp cli 
+RUN curl "$GOOGLEURL" -o "googlecloudsdk.tar.gz"
+RUN tar -xvzf googlecloudsdk.tar.gz; sleep 1
+RUN rm googlecloudsdk.tar.gz
+ENV PATH="$PATH:/google-cloud-sdk/bin"
+
+# terraform
+RUN curl "$TERRAFORMURL" -o "terraform.zip"
+RUN unzip terraform.zip -d /usr/local/bin
+RUN chmod +x /usr/local/bin/terraform
+RUN rm terraform.zip
+
+# kubectl
+RUN curl -LO "$KUBECTLURL"
+RUN chmod +x kubectl
+RUN mv kubectl /usr/local/bin
+
+# helm - script is smart enough to handle different architectures
+RUN curl https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3 | bash
+
+# locales
+RUN apt-get install -y locales locales-all
+ENV LC_ALL en_US.UTF-8
+ENV LANG en_US.UTF-8
+ENV LANGUAGE en_US.UTF-8
+ENV DOCKER_HOST=tcp://localhost:2375
+
+
+# build time env args
+
+ARG HOME
+ARG OPTAVERSION
+
+# add user
+RUN groupadd -g ${CONTAINERGID} ${USER}
+RUN useradd ${USER} -m -d ${HOME} -u ${MYUID} -g ${CONTAINERGID} 
+
+# docker (install after adding docker group above)
+RUN curl $DOCKERURL -o docker.tar.gz
+RUN tar xvzf docker.tar.gz
+RUN cp docker/* /usr/local/bin
+RUN rm docker.tar.gz
+RUN rm -rf docker/
+
+# clean up
+RUN apt-get clean
+
+# opta
+ADD install.sh /tmp/install.sh
+RUN /bin/bash -c "/tmp/install.sh"
+
+
+USER ${USER}
+WORKDIR ${HOME}
+
+ENTRYPOINT ["/usr/local/bin/opta/opta"]
+EOF
+
+echo "Building your Opta image"
+# Build the docker image
+docker build --quiet /tmp/opta \
```

Comment:
We should build the docker images and push them to docker hub - instead of each user building them.

Reply:
This was infact the original plan.

The issue is that if we make the image and put it in dockerhub for the user to pull, then we will need the user to run the docker container as a privileged root user (inside the docker container); which means that any files we touch with opta have root:root permission on the host as well. For example, the `.opta/local` directory and its contents, and possibly other files opta executable touches...

In the current opta.sh script, the user is detected and a user with the identical uid and gid is created (for Linux).

For Mac, since the host gid is < 1000 we cant do this (because the Linux based container we are running with opta.sh will not allow users inside the running container to have gids <1000, so gid is set to 12345.